// @vuepress-reco/vuepress-plugin-comments/lib/client/components/Valine.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const vue_1 = require("vue");
const vue_router_1 = require("vue-router");
require("../styles/valine.css");
exports.default = vue_1.defineComponent({
    name: 'Valine',
    props: {
        options: {
            type: Object,
            default() {
                return {};
            }
        }
    },
    setup(props) {
        const route = vue_router_1.useRoute();
        const { options } = vue_1.toRefs(props);
        const initValine = () => {
            const Valine = require('valine');
            const valineOptions = {
                el: '#valine',
                placeholder: 'just go go',
                notify: false,
                verify: false,
                avatar: 'retro',
                visitor: true,
                recordIP: false,
                path: window.location.pathname,
                ...options.value
            };
            new Valine(valineOptions);
        };
        vue_1.onMounted(() => {
            initValine();
        });
        return { initValine };
    },
    render() {
        return vue_1.h('div', {
            class: 'reco-valine-wrapper'
        }, vue_1.h('div', {
            id: 'valine'
        }));
    },
    watch: {
        '$route'(from, to) {
            if (to.path !== from.path) {
                // 切换页面时刷新评论
                setTimeout(() => {
                    this.initValine();
                }, 300);
            }
        }
    }
});
