// @vuepress-reco/vuepress-plugin-comments/lib/client/components/ValineViews.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const vue_1 = require("vue");
const client_1 = require("@vuepress/client");
const vue_router_1 = require("vue-router");
exports.default = vue_1.defineComponent({
    name: 'ValineViews',
    props: {
        idVal: String,
        numStyle: Object,
        flagTitle: {
            type: String,
            default: 'Your Article Title'
        }
    },
    setup(props) {
        const siteDate = client_1.useSiteData();
        const route = vue_router_1.useRoute();
        const { idVal, numStyle, flagTitle } = vue_1.toRefs(props);
        const getIdVal = (path) => {
            return siteDate.value.base.slice(0, siteDate.value.base.length - 1) + path;
        };
        return () => vue_1.h('span', {
            id: getIdVal(idVal.value || route.path),
            class: 'leancloud-visitors',
            'data-flag-title': flagTitle.value
        }, vue_1.h('a', {
            class: 'leancloud-visitors-count',
            style: numStyle.value
        }));
    }
});
