// @vuepress-reco/vuepress-plugin-comments/lib/client/components/Vssue.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const vue_1 = require("vue");
const vue_router_1 = require("vue-router");
const vssue_1 = require("vssue");
const api_github_v3_1 = require("@vssue/api-github-v3");
const api_github_v4_1 = require("@vssue/api-github-v4");
const api_gitlab_v4_1 = require("@vssue/api-gitlab-v4");
const api_bitbucket_v2_1 = require("@vssue/api-bitbucket-v2");
const api_gitee_v5_1 = require("@vssue/api-gitee-v5");
require("vssue/dist/vssue.css");
exports.default = vue_1.defineComponent({
    name: 'Vssue',
    props: {
        options: {
            type: Object,
            default() {
                return {};
            }
        }
    },
    setup(props) {
        const route = vue_router_1.useRoute();
        const { options } = vue_1.toRefs(props);
        const key = vue_1.ref('key');
        const platformOptions = {
            'github': api_github_v3_1.default,
            'github-v4': api_github_v4_1.default,
            'gitlab': api_gitlab_v4_1.default,
            'bitbucket': api_bitbucket_v2_1.default,
            'gitee': api_gitee_v5_1.default
        };
        const vssueOptions = vue_1.computed(() => {
            const platform = platformOptions[options.value.platform || 'github'];
            return { ...options.value, api: platform };
        });
        vue_1.onMounted(() => {
            key.value = `reco-${new Date().getTime()}`;
        });
        vue_1.watch(route, (to, from) => {
            if (to.path !== from.path) {
                // 切换页面时刷新评论
                setTimeout(() => {
                    key.value = `reco-${new Date().getTime()}`;
                }, 300);
            }
        });
        return () => vue_1.h(vssue_1.VssueComponent, {
            class: 'reco-vssue-wrapper',
            options: vssueOptions.value
        });
    }
});
